WIP: Render prompt asyncronuosly using zsh-async

salmanulfarzy opened the pull request:
#315 already exists for async rendering, Didn't want to mess up @sei40kr work there by pushing into it so made a new PR. To see previous discussion on async see #307 and #315  

- [x] Loading `zsh-async` (>= `v1.7.0`)
- [x] API for sections
- [ ] API documentation
- [ ] Async `git` section
   - [ ] async `git_branch` section
   - [ ] async `git_status` section
- [ ] Updated testcases

maximbaz commented on spaceship.zsh:
```diff
@@ -179,7 +179,11 @@ spaceship_async_callback() {
+  if spaceship::exists async; then
+    autoload -Uz async && async
+  else
+    source $SPACESHIP_ROOT/async.zsh && async_init
```
use `&& async_init` or `&& async` in both `if` and `else`, for consistency 😛

maximbaz commented on spaceship.zsh:
```diff
@@ -178,6 +179,11 @@ spaceship_ps2() {
+  if spaceship::exists async; then
+    autoload -Uz async && async
+  else
+    source $SPACESHIP_ROOT/async.zsh && async_init
```
use `&& async_init` or `&& async` in both `if` and `else`, for consistency 😛

salmanulfarzy replied:
`async_init` is [required only](https://github.com/mafredri/zsh-async#async_init) if `async` unavailable on `$fpath`.

maximbaz replied:
I think `async` is just a beautiful alias over `async_init`, we can use one or another regardless of how it was loaded:

https://github.com/mafredri/zsh-async/blob/5248bd85bf2c522ebfa2b8d18ee7bc736d10c791/async.zsh#L543-L545

maximbaz commented:
Looks good for a start 👍

salmanulfarzy commented:
> Sometimes random numbers popping up on `git_status` section

These were specifically related to our `git_status` section and it's indicators.

- Untracked files ➜  Exit code of previous command  ([image](https://user-images.githubusercontent.com/10276208/44300815-9fd09100-a32a-11e8-8aca-d18618129624.png))
- Modified unstaged files ➜  Random numbers, Yet to figure out what it is.

**Correct behavior** on

- Stashes
- Staged files
- Deleted
- Rename
- Modified + Renamed + Unstaged

Last one was surprise for me and needs for information.

laggardkernel commented:
It seems the async implementation breaks the vi_mode section. vi_mode section doesn't change its content anymore.

salmanulfarzy commented:
I'm busy with work now. Don't think would be able to get back on this soon. Feel free if someone want to work on this or takeover.

laggardkernel commented:
The async worker should be restarted in `precmd` to refresh the shell environment. Otherwise, new `export`ed environ variables are not available in async job.

https://github.com/robobenklein/zinc/blob/28bea2b33e8fe1fc1319eab3c034f81bec8601b5/zinc_functions/prompt_zinc_setup#L102-L109

```zsh
# restarts just the worker - in order to update worker with current shell values
function zinc_worker_restart () {
  # _ZINC_DBG_OUT "restarting worker..."
  # async_worker_eval zinc_segment_worker "zinc_reload"
  async_stop_worker zinc_segment_worker
  async_start_worker zinc_segment_worker
  async_register_callback zinc_segment_worker _zinc_segment_async_callback
}

maximbaz commented:
I don't think that's needed, but I can't remember details at the moment - check https://github.com/mafredri/zsh-async/pull/29 and the linked issue to that PR.

laggardkernel commented:
@maximbaz Sorry, maybe I didn't make it clear. The async worker doesn't have access to any variables (not only `export` type) **set later** after the worker starts up. It will keep using the **stale shell environment** when it's set up. Besides, some unset variables may be still available in the stale environment. That's the reason we need to restart the async worker at `precmd`.

`async_worker_eval` is used to change specific variables manually inside the worker, which is not enough to refresh all the environment variables.

maximbaz commented:
Ah, I see now, thanks! So is the main purpose of restarting worker all the time to allow configuring prompt in runtime (e.g. by executing `export SPACESHIP_EXEC_TIME_SHOW=false` directly in the terminal), as opposed to restarting the terminal once you made your changes to the config file? It is definitely a valid scenario, I simply want to understand if there's more to it - I don't think prompt cares about any other environment variables, right?

laggardkernel commented:
> So is the main purpose of restarting worker all the time to allow configuring prompt in runtime ...

@maximbaz No, the main purpose of refreshing environment of the worker is to make the **`async_job`s** (used by **async plugins** and run by the worker) work properly.

I'll take `pyenv`, the python version manager, for an example. (The following process applies to `rbenv`, `nodenv` as well.) Using `pyenv`, you can set a specific python version or virtual environment used _under the current dir_ with one of the following methods,
- `export PYENV_VERSION=<venv-name>` manually
- Use `pyenv local <venv-name>` to write `<venv-name>` in the `.python-version` file under current working dir

The command `pyenv version-name`, which is used in the `pyenv ` plugin, will try to retrieve the current python version from the places above.

If you make the `pyenv version-name` as an `async_job`, the plugin doesn't work as expected. Cause the `PYENV_VERSION` may be stale or not available at all in the worker with a stale shell environment.

I notice there's an async `ruby` plugin in your fork, you can replace the word `pyenv` with `rbenv` in the above words to understand the whole process.

maximbaz commented:
I see, thanks for the explanation!

raihan26 commented:
any update on this PR? Seeing my prompt very slow.

loganbek commented:
Hey what’s up? Can i be of help to you?

On Sun, May 3, 2020 at 3:21 PM raihan26 <notifications@github.com> wrote:

> any update on this PR? Seeing my prompt very slow.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/denysdovhan/spaceship-prompt/pull/499#issuecomment-623191699>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACHL57DZI7FQSQZBHSGRHL3RPXU5PANCNFSM4FQKPIQA>
> .
>
-- 
Logan Bek
LinkedIn <http://www.linkedin.com/in/loganbek>
Github <https://github.com/loganbek>
Come work for the open internet w/ Gitcoin
<https://gitcoin.co/townsquare/?cb=ref:e727>
loganbek@gmail.com

denysdovhan commented:
I'd rather keep this open if you don't mind, but with corresponding badge.

salmanulfarzy commented:
Sure. I'll leave this PR open and branch unmodified.
